File: src/analyze_includes/result.py
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field
from json import dumps
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pathlib import Path

    from src.analyze_includes.parse_source import Include


@dataclass
class Result:
    target: str
    report: Path | None = None
    use_impl_deps: bool = False

    public_includes_without_dep: list[Include] = field(default_factory=list)
    private_includes_without_dep: list[Include] = field(default_factory=list)
    unused_deps: list[str] = field(default_factory=list)
    unused_impl_deps: list[str] = field(default_factory=list)
    deps_which_should_be_private: list[str] = field(default_factory=list)

    def is_ok(self) -> bool:
        return (
            len(self.public_includes_without_dep) == 0
            and len(self.private_includes_without_dep) == 0
            and len(self.unused_deps) == 0
            and len(self.unused_impl_deps) == 0
            and len(self.deps_which_should_be_private) == 0
        )

    def to_str(self) -> str:
        msg = f"DWYU analyzing: '{self.target}'\n\n"
        if self.is_ok():
            msg += "Result: SUCCESS"
            return self._framed_msg(msg)

        msg += "Result: FAILURE\n"
        if self.public_includes_without_dep or self.private_includes_without_dep:
            msg += "\nIncludes which are not available from the direct dependencies:\n"
            msg += "\n".join(f"  {inc}" for inc in self.public_includes_without_dep + self.private_includes_without_dep)
        if self.unused_deps:
            msg += "\nUnused dependencies in 'deps' (none of their headers are referenced):\n"
            msg += "\n".join(f"  Dependency='{dep}'" for dep in self.unused_deps)
        if self.unused_impl_deps:
            msg += "\nUnused dependencies in 'implementation_deps' (none of their headers are referenced):\n"
            msg += "\n".join(f"  Dependency='{dep}'" for dep in self.unused_impl_deps)
        if self.deps_which_should_be_private:
            msg += "\nPublic dependencies which are used only in private code:\n"
            msg += "\n".join(f"  Dependency='{dep}'" for dep in self.deps_which_should_be_private)

        msg += f"\n\nDWYU Report: {self.report}"

        return self._framed_msg(msg)

    def to_json(self) -> str:
        content = {
            "analyzed_target": self.target,
            "public_includes_without_dep": self._make_includes_map(self.public_includes_without_dep),
            "private_includes_without_dep": self._make_includes_map(self.private_includes_without_dep),
            "unused_deps": self.unused_deps,
            "unused_implementation_deps": self.unused_impl_deps,
            "deps_which_should_be_private": self.deps_which_should_be_private,
            "use_implementation_deps": self.use_impl_deps,
        }
        return dumps(content, indent=2) + "\n"

    @staticmethod
    def _make_includes_map(includes: list[Include]) -> defaultdict[str, list[str]]:
        includes_mapping = defaultdict(list)
        for inc in includes:
            includes_mapping[str(inc.file)].append(inc.include)
        return includes_mapping

    @staticmethod
    def _framed_msg(msg: str) -> str:
        """Put a message between 2 horizontal borders"""
        border = 80 * "="
        return border + "\n" + msg + "\n" + border

File: dwyu/aspect/private/analyze_includes/result.cpp
#include "dwyu/aspect/private/analyze_includes/result.h"

#include <nlohmann/json.hpp>

#include <fstream>
#include <string>
#include <vector>

using json = nlohmann::json;

namespace dwyu {
namespace {

std::map<std::string, std::vector<std::string>> makeMissingIncludesMap(const std::vector<IncludeStatement>& includes) {
    std::map<std::string, std::vector<std::string>> map{};
    for (const auto& inc : includes) {
        const auto it_file = map.find(inc.file);
        if (it_file == map.end()) {
            map.insert({inc.file, {inc.include}});
        }
        else {
            it_file->second.push_back(inc.include);
        }
    }
    return map;
}

std::string removeQuoting(const std::string& str) {
    if (str.size() >= 2) {
        if ((str.front() == '"' && str.back() == '"') || (str.front() == '<' && str.back() == '>')) {
            return str.substr(1, str.size() - 2);
        }
    }
    return str;
}

} // namespace

Result::Result(const std::string& target, bool optimize_impl_deps)
    : target_{target}, optimize_impl_deps_{optimize_impl_deps} {}

bool Result::isOk() const {
    return public_includes_without_direct_dep_.empty() && private_includes_without_direct_dep_.empty() &&
           unused_deps_.empty() && unused_impl_deps_.empty() && public_dep_which_should_be_private_.empty();
}

void Result::setPublicIncludesWithoutDirectDep(std::vector<IncludeStatement> includes) {
    public_includes_without_direct_dep_ = std::move(includes);
}

void Result::Result::setPrivateIncludesWithoutDirectDep(std::vector<IncludeStatement> includes) {
    private_includes_without_direct_dep_ = std::move(includes);
}

void Result::setUnusedDeps(std::vector<std::string> deps) {
    unused_deps_ = std::move(deps);
}

void Result::setUnusedImplDeps(std::vector<std::string> deps) {
    unused_impl_deps_ = std::move(deps);
}

void Result::setDepsWhichShouldBePrivate(std::vector<std::string> deps) {
    public_dep_which_should_be_private_ = std::move(deps);
}

std::string Result::toString(const std::string& report_path) const {
    const std::string frame{"================================================================================\n"};

    std::string content{"DWYU analyzing: '" + target_ + "'\n\n"};
    if (isOk()) {
        content += "Result: SUCCESS\n";
        return frame + content + frame;
    }

    content += "Result: FAILURE\n";

    if (!public_includes_without_direct_dep_.empty() || !private_includes_without_direct_dep_.empty()) {
        content += "\nIncludes which are not available from the direct dependencies:\n";
        for (const auto& include : public_includes_without_direct_dep_) {
            // TODO change to 'content += "  In file '" + include.file + "' include: " + include.include + "\n";'
            // whenPython impl is gone
            content += "  File='" + include.file + "', include='" + removeQuoting(include.include) + "'\n";
        }
        for (const auto& include : private_includes_without_direct_dep_) {
            // TODO change to 'content += "  In file '" + include.file + "' include: " + include.include + "\n"'
            // whenPython impl is gone
            content += "  File='" + include.file + "', include='" + removeQuoting(include.include) + "'\n";
        }
    }

    if (!unused_deps_.empty()) {
        // TODO change to 'Unused dependencies in 'deps' (none of their headers are included):' whenPython impl is gone
        content += "\nUnused dependencies in 'deps' (none of their headers are referenced):\n";
        for (const auto& dep : unused_deps_) {
            content += "  " + dep + "\n";
        }
    }

    if (!unused_impl_deps_.empty()) {
        // TODO change to 'Unused dependencies in 'implementation_deps' (none of their headers are included):'
        // whenPython impl is gone
        content += "\nUnused dependencies in 'implementation_deps' (none of their headers are referenced):\n";
        for (const auto& dep : unused_impl_deps_) {
            // TODO simplify to 'content += "  " + dep + "\n";' when Python impl is gone
            content += "  Dependency='" + dep + "'\n";
        }
    }

    if (!public_dep_which_should_be_private_.empty()) {
        // TODO change to 'Dependencies in 'deps' used only in private code which should be moved to
        // 'implementation_deps':' whenPython impl is gone
        content += "\nPublic dependencies which are used only in private code:\n";
        for (const auto& dep : public_dep_which_should_be_private_) {
            // TODO simplify to 'content += "  " + dep + "\n";' when Python impl is gone
            content += "  Dependency='" + dep + "'\n";
        }
    }

    content += "\n";
    content += "DWYU Report: " + report_path + "\n";

    return frame + content + frame;
}

nlohmann::json Result::toJson() const {
    json data{};
    data["analyzed_target"] = target_;
    data["public_includes_without_dep"] = makeMissingIncludesMap(public_includes_without_direct_dep_);
    data["private_includes_without_dep"] = makeMissingIncludesMap(private_includes_without_direct_dep_);
    data["unused_deps"] = unused_deps_;
    data["unused_implementation_deps"] = unused_impl_deps_;
    data["deps_which_should_be_private"] = public_dep_which_should_be_private_;
    data["use_implementation_deps"] = optimize_impl_deps_;
    return data;
}

} // namespace dwyu
